update config document (#1751)
vdaas/vald: docs/user-guides/configuration.md

The logging level is defined in [the Coding Style Guide](../contributing/coding-style.md#logging).

### Servers

Each Vald component has several types of servers.
They can be configured by specifying the values in `defaults.server_config`.

> In addThey can be overwritten by specifying `[component].server_config`.

Reviewer: ```suggestion
In addition, they can be overwritten by specifying `[component].server_config`.
```

            ...
```

#### gRPC server

gRPC server should be enabled because all Vald components use gRPC to communicate with others.
The API specs are placed in [Vald APIs](../api).

#### REST server

REST server is optional.
The swagger specs are placed in [Vald APIs Swagger][vald-swagger-specs].

#### Health check servers

There are two built-in health check servers: liveness and readiness.
They are used as servers for [Kubernetes liveness and readiness probe][kubernetes-liveness-readiness].

> Liveness servers are default disabled for the agent because the liveness probes may accidentally kill it.

Reviewer 2: ```suggestion
The liveness health server is disabled by default due to the liveness probe may accidentally kill the Agent component.
```

        enabled: false
```

### Metrics servers

> Metrics servers help to debug and monitor Vald components.

Reviewer 2: ```suggestion
The metrics server enables easier debugging and monitoring of Vald components.
```

There are two types of metrics servers: pprof and Prometheus.

pprof server is implemented using Go's `net/http/pprof` package.

> You can use [google's pprof][google-pprof] to analyze the exported profiling data.

Reviewer 2: ```suggestion
You can use [google's pprof][google-pprof] to analyze the exported profile result.
```

Prometheus server is a [Prometheus][prometheus-io] exporter.
It is required to set the `observability` section on each Vald component to enable the monitoring using Prometheus.
Please refer to the next section.

### Observability

The observability features are useful for monitoring Vald components.

> They can be enabled by setting the value `true` on the `defaults.observability.enabled` field or overriding it in each component (`[component].observability.enabled`).

Reviewer 2: ```suggestion
These settings can be enabled by setting the `defaults.observability.enabled` field to the value `true` or by overriding it in each component (`[component].observability.enabled`).
```

        enabled: false
```

### Metrics servers

> Metrics servers help debug and monitor Vald components.

Reviewer: ```suggestion
Metrics servers help to debug and monitor Vald components.
```

- `agent.ngt.distance_type`
- `agent.ngt.object_type`

Users should configure these parameters first to fit their use case.

For further details, please read [the NGT wiki][yj-ngt-wiki].

Vald Agent NGT has a feature to start indexing automatically.
The behavior of this feature can be configured with these parameters:

- `agent.ngt.auto_index_duration_limit`
- `agent.ngt.auto_index_check_duration`
- `agent.ngt.auto_index_length`

<div class="notice">

> When Vald Agent NGT is indexing, all search requests to the target pod will be canceled.

Reviewer 2: ```suggestion
While the Vald Agent NGT is in the process of creating indexes, it will ignore all search requests to the target pods.
```

</div>

<div class="warning">

> When deploying Vald Index Manager, the above parameters should be set much longer than the Vald Index Manager settings.

Reviewer 2: A much longer number is sensory intellectually difficult.
User will confuse about the number how long is enought or not.

> When deploying Vald Index Manager, the above parameters should be set much longer than the Vald Index Manager settings, e.g., set agent.ngt.auto_index_duration_limit to "720h" and agent.ngt.auto_index_check_duration to "24h".

Author: ```suggestion
When deploying Vald Index Manager, the above parameters should be set much longer than the Vald Index Manager settings, e.g., designated agent.ngt.auto_index_duration_limit to "720h" and agent.ngt.auto_index_check_duration to "24h".
```

> When deploying Vald Index Manager, the above parameters should be set much longer than the Vald Index Manager settings, e.g., designated agent.ngt.auto_index_duration_limit to "720h" and agent.ngt.auto_index_check_duration to "24h".

Reviewer: ```suggestion
When deploying Vald Index Manager, the above parameters should be set much longer than the Vald Index Manager settings, e.g., setting the value of agent.ngt.auto_index_duration_limit to "720h" and agent.ngt.auto_index_check_duration to "24h".
```

- `agent.ngt.distance_type`
- `agent.ngt.object_type`

Users should configure these parameters first to fit their use case.

For further details, please read [the NGT wiki][yj-ngt-wiki].

Vald Agent NGT has a feature to start indexing automatically.
The behavior of this feature can be configured with these parameters:

- `agent.ngt.auto_index_duration_limit`
- `agent.ngt.auto_index_check_duration`
- `agent.ngt.auto_index_length`

<div class="notice">
When Vald Agent NGT is indexing, all search requests to the target pod will be canceled.
</div>

<div class="warning">
When deploying Vald Index Manager, the above parameters should be set much longer than the Vald Index Manager settings.
(Please see Vald Index Manager section)<BR>

> This is because Vald Index Manager accurately grasps the index information of each Vald Agent NGT and controls the execution timing of indexing.<BR><BR>

Reviewer: ```suggestion
This is because the Vald Index Manager accurately grasps the index information of each Vald Agent NGT and controls the execution timing of indexing.<BR><BR>
```

When the setting parameter of Vald Agent NGT is shorter than the setting value of Vald Index Manager, Vald Agent NGT may start indexing by itself without the execution command from Vald Index Manager.
If this happens, the Index Manager may not function properly.
</div>

#### Resource requests and limits, Pod priorities

> Because the agent places indices on memory, termination of agent pods causes loss of indices.

Reviewer: ```suggestion
The agent places indexes on memory, termination of agent pods causes loss of indexes.
```

Author: `Because` is needed.
But, I will replace `indices` to `indexes`

> It is important to set resource requests and limits appropriately not to terminate agent pods.

Reviewer: ```suggestion
It is important to set the resource request and limit appropriately to avoid terminating the agent pods.
```

Requesting 40% of cluster memory for agent pods is highly recommended.

> And also, it is highly recommended not to set resource limits to agent pods.

Reviewer: ```suggestion
Also, it is highly recommended not to set resource limits to agent pods.
```

#### Index replica

`gateway.lb.gateway_config.index_replica` represents how many Vald Agent pods a vector will be inserted into.

> We recommend setting it as one-third of the number of Vald Agent pods.

Reviewer: I think we didn't recommend to set it to one-third.
e.g. User deployed 30 agents, in this sentence it means we recommended to set the index replica to 10.

```suggestion
We recommend the maximum value of the index replica is one-third of the Vald agent deployed.
```

Author: ```suggestion
The maximum value of the index replica should be one-third of the Vald Agent pods deployed.
```

Author: @kevindiu how about it?

Reviewer: @vankichi LGTM

```yaml
gateway:
  lb:
    gateway_config:

>       index_replica: 3 // By setting the index replica to 3, the number of Vald Agent pods deployed should be more than 9.

Reviewer 2: I think it would be better to describe value as a percentage.
in this case it would be 30%

#### Index replica

`gateway.lb.gateway_config.index_replica` represents how many Vald Agent pods a vector will be inserted into.
We recommend setting it as one-third of the number of Vald Agent pods.

```yaml
gateway:
  lb:
    gateway_config:

>       index_replica: 3 // The number of Vald Agent pods should be larger than 9.

Reviewer: ```suggestion
      index_replica: 3 // By setting the index replica to 3, the number of Vald Agent pods deployed should be more than 9.
```

### Discoverer

#### Cluster Role

> Vald Discoverer gets the Node and Por metrics from [kube-apiserver](https://kubernetes.io/ja/docs/reference/command-line-tools-reference/kube-apiserver/) as described in [Vald Discoverer](../overview/component/discoverer.md).

Reviewer: ```suggestion
Vald Discoverer gets the Node and Pod metrics from [kube-apiserver](https://kubernetes.io/ja/docs/reference/command-line-tools-reference/kube-apiserver/) as described in [Vald Discoverer](../overview/component/discoverer.md).
```

Vald's Helm deployment supports RBAC as default, and the default configuration is the following.

```yaml
discoverer:
  clusterRole:
    # if true, the clusterRole configuration will be created.
    enabled: true
    name: discoverer
  clusterRoleBinding:
    # if true, the clusterRoleBinding configuration will be created.
    enabled: true
    name: discoverer
  serviceAccount:
    # if true, the serviceAccount configuration will be created.
    enabled: true
    name: vald
```

When `RBAC` is unavailable in your environment, or you would like to put some restrictions, please modify it and grant the permissions to the user executing the discoverer.
Each configuration file is the following:

> - [cluster role](https://github.com/vdaas/vald/blob/master/k8s/discoverer/clusterrole.yaml)
> - [cluster role binding](https://github.com/vdaas/vald/blob/master/k8s/discoverer/clusterrolebinding.yaml)
> - [service account](https://github.com/vdaas/vald/blob/master/k8s/discoverer/serviceaccount.yaml)

Reviewer: ```suggestion
- [Cluster role](https://github.com/vdaas/vald/blob/master/k8s/discoverer/clusterrole.yaml)
- [Cluster role binding](https://github.com/vdaas/vald/blob/master/k8s/discoverer/clusterrolebinding.yaml)
- [Service account](https://github.com/vdaas/vald/blob/master/k8s/discoverer/serviceaccount.yaml)
```

#### Resource requests and limits

The number of discoverer pods and resource limits can be estimated by the configurations of your LB gateways and index managers because its APIs are called by them.
Discoverer CPU loads almost depend on API request traffic = (the number of LB gateways x its request frequency) + (the number of index managers x its request frequency).

### Index Manager

#### Execution index command to Vald Agent

Vald Index Manager controls the indexing timing for all Vald Agent pods in the Vald cluster.

> These parameters relate control process.

Reviewer: ```suggestion
These parameters are related to the control process.
```

```yaml
manager:
  index:
    indexer:
      # namespace of agent pods to manage
      agent_namespace: vald # namespace of agent pods to manage
      # check duration of automatic indexing
      auto_index_check_duration: "1m"
      # limit duration of automatic indexing
      auto_index_duration_limit: "30m"
      # number of caches to trigger automatic indexing
      auto_index_length: 100
      # limit duration of automatic index saving
      auto_save_index_duration_limit: "3h"
      # duration of automatic index saving wait duration for next saving
      auto_save_index_wait_duration: "10m"
      # the number of Agent Pods indexing at the same time
      concurrency: 1
      # number of pool size of creating index processing
      creation_pool_size: 10000
```

#### Discoverer request duration

Same as LB gateway, manager.index.indexer.discoverer.duration represents the frequency of sending requests to the discoverer.

## References

> For further details, there are references to Helm values in the GitHub Vald repository.

Reviewer: ```suggestion
For further details, there are references to the Helm values in the Vald GitHub repository.
```
